### DISCUSSION TITLE
Please change Accept-Encoding for requests to /versions for rubygems servers

### DISCUSSION THREAD

**Discussion by segiddins** [renovatebot/renovate#21881] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Hi there, I'm one of the maintainers of rubygems.org.

Recently, we've been experiencing some performance issues with the service, so I've spent a bunch of time investigating possible causes. One thing that has stood out has been the amount of time we've been spending generating responses to /versions requests.

After digging into our metrics, I've discovered that almost all of the cache misses are coming when `Accept-Encoding` do not include `gzip`, and almost all of those requests are coming from Renovate.

Changing to `Accept-Encoding: gzip` would save rubygems.org something like 3 days of compute per day, as all those cache misses would likely become hits served by our CDN.

**Comment by segiddins** [comment]
(https://github.com/renovatebot/renovate/blob/d166bc01e64bc8c1ef28aaebe2bea88195c9c801/lib/modules/datasource/rubygems/versions-datasource.ts#L76 is the line that would have to be changed)

**Comment by rarkins** [comment]
Definitely! @zharinov has already been looking into this code in the last day but we weren't aware of the gzip aspect. We'll try to do that right away

**Comment by zharinov** [comment]
Also, we've found that we actually set `range: bytes=0-` for each request, instead of calculating the offset and fetching the delta. Is it okay from rubygems performance standpoint, if we fix it and if we'll use `range: bytes=N-` with different values of `N`?

**Comment by segiddins** [comment]
Ah yeah definitely!On Apr 28, 2023, at 10:21 PM, Sergei Zharinov ***@***.***> wrote:﻿
Also, we've found that we actually set range: bytes=0- for each request, instead of calculating the offset and fetching the delta. Is it okay from rubygems performance standpoint, if we fix it and if we'll use range: bytes=N- with different values of N?

**Comment by HonkingGoose** [comment]
I created this issue to track the work:

- #21887

**Comment by zharinov** [comment]
Our HTTP client fails when compression set to `gzip` and the `range` header is like `bytes=12-31`. However, seems like it works fine with `deflate, compress, br`.

@segiddins Could you recommend one of these 3 that would work well for rubygems too?

**Reply by segiddins** [reply]
I think you might need to leave out the accept-encoding when passing range? just make sure that you're not making `range: bytes=0-` requests, because an un-ranged, compressed request would be preferable to that

**Reply by zharinov** [reply]
Seems like omitted `accept-encoding` defaults to `gzip`. I think we'll set `gzip` for unranged fetch-all requests, while for ranged ones we'll set it to `deflate, compress, br, identity`.
